pennylane/qnode.py

```diff
@@ -567,7 +566,8 @@ def __call__(self, *args, **kwargs):
             **self.execute_kwargs,
         )
 
-        res = processing_fn(res)
+        if autograd.isinstance(res, (tuple, list)) and len(res) == 1:
+            res = res[0]

```

Comment:
I vaguely know why this is needed --- previously, the batch `processing_fn` acted to convert `list[float] -> float` if a batch transform was applied. If no batch transform was applied, then nothing happened.

However, now we have a dilemma: how do we know _when_ a batch transform has been applied, so that we can map `list[float] -> float` as before?

There must be a better way of doing this; my current approach is simply trial and error by exploring test failures.

Reply:
Is it ok to use autograd here? Could `res` be in a different interface?

Reply:
Also, could the answer lie within the device's `batch_transform` method, i.e., we just need to mark if we map to multiple tapes there?

Reply:
> Is it ok to use autograd here? Could res be in a different interface?

This is a very good question :) This was originally simply

```python
isinstance(res, (tuple, list))
```

This worked for both forward and backwards passes of TF, Torch, and JAX, but was only working on the forwards pass for Autograd.

I dug deeper, and discovered that on the backwards pass, Autograd converts lists and tuples to `autograd.builtins.SequenceBox` objects, so they wouldn't be detected by this check.

So there are two solutions; firstly I tried

```python
isinstance(res, (tuple, list, autograd.builtins.list, autograd.builtins.tuple))
```

This worked, but I thought seemed a bit too strange. I dug through the Autograd source code, and discovered that they provide a slightly modified `isinstance` function that works on Autograd backwards passes, `autograd.isinstance`. I swapped it in here, and then this check worked on all interfaces :)

I'll add a comment explaining this!

```suggestion
        if autograd.isinstance(res, (tuple, list)) and len(res) == 1:
            # If a device batch transform was applied, we need to 'unpack'
            # the returned tuple/list to a float.
            #
            # Note that we use autograd.isinstance, because on the backwards pass
            # with Autograd, lists and tuples are converted to autograd.box.SequenceBox.
            # autograd.isinstance is a 'safer' isinstance check that supports
            # autograd backwards passes.

            res = res[0]
```

> Also, could the answer lie within the device's batch_transform method, i.e., we just need to mark if we map to multiple tapes there?

Yeah that's a good suggestion. I spent a while thinking about this, and I can't work out how to communicate this back to `qml.execute` and then back to the `qnode`, without having to change the return format of `qml.execute` 🤔 Do you have any ideas?